File: Using TDD with ASP.NET MVC/[C#]-Using TDD with ASP.NET MVC/C#/MvcContacts.Tests/Models/InMemoryContactRepository.cs
// <snippet5>
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using MvcContacts.Models;

namespace MvcContacts.Tests.Models {
    class InMemoryContactRepository : MvcContacts.Models.IContactRepository {
        private List<Contact> _db = new List<Contact>();

        public Exception ExceptionToThrow { get; set; }
        //public List<Contact> Items { get; set; }

        public void SaveChanges(Contact contactToUpdate) {

            foreach (Contact contact in _db) {
                if (contact.Id == contactToUpdate.Id) {
                    _db.Remove(contact);
                    _db.Add(contactToUpdate);
                    break;
                }
            }
        }

        public void Add(Contact contactToAdd) {
            _db.Add(contactToAdd);
        }

        public Contact GetContactByID(int id) {
            return _db.FirstOrDefault(d => d.Id == id);
        }

        public void CreateNewContact(Contact contactToCreate) {
            if (ExceptionToThrow != null)
                throw ExceptionToThrow;

            _db.Add(contactToCreate);
           // return contactToCreate;
        }

        public int SaveChanges() {
            return 1;
        }

        public IEnumerable<Contact> GetAllContacts() {
            return _db.ToList();
        }


        public void DeleteContact(int id) {
            _db.Remove(GetContactByID(id));
        }

    }
} 
// </snippet5>

File: Using TDD with ASP.NET MVC/[VB]-Using TDD with ASP.NET MVC/VB/MvcContacts.Tests/Models/InMemoryContactRepository.vb
' <snippet5>

Class InMemoryContactRepository
    Implements IContactRepository

    Private _db As New List(Of Contact)()
    Public Property ExceptionToThrow() As Exception

    Public Function SaveChanges() As Integer Implements IContactRepository.SaveChanges
        Return 1
    End Function

    Public Sub Add(ByVal contactToAdd As Contact)
        _db.Add(contactToAdd)
    End Sub

    Public Function GetContactByID(ByVal id As Integer) As Contact Implements IContactRepository.GetContactByID
        Return _db.FirstOrDefault(Function(d) d.Id = id)
    End Function

    Public Sub CreateNewContact(ByVal contactToCreate As Contact) Implements IContactRepository.CreateNewContact
        If ExceptionToThrow IsNot Nothing Then
            Throw ExceptionToThrow
        End If

        _db.Add(contactToCreate)
    End Sub


    Public Function GetAllContacts() As IEnumerable(Of Contact) Implements IContactRepository.GetAllContacts
        Return _db.ToList()
    End Function


    Public Sub DeleteContact(ByVal id As Integer) Implements IContactRepository.DeleteContact
        _db.Remove(GetContactByID(id))
    End Sub

End Class

' </snippet5>
